app/vmselect/prometheus/prometheus.go

```diff
@@ -901,6 +938,11 @@ func QueryHandler(qt *querytracer.Tracer, startTime time.Time, at *auth.Token, w
 		DenyPartialResponse: httputils.GetDenyPartialResponse(r),
 		QueryStats:          qs,
 	}
+	err = populateAuthTokens(qt, ec, at, deadline)
```

Comment:
It's better to parse tenant related labels only once, during query request to the storage. 

 It ll make code simpler and allow to use current cache implementation.